src/PairedUrl.php

```diff
@@ -0,0 +1,146 @@
+<?php
```

Comment:
This more looks like a helper trait than a service, so I'd recommend making it one.

Reply:
Or else it the methods in that service could be just made methods on the abstract class directly. 

The reason a service was chosen was based on the advice of @schlessera. 

Also, the service is being used not only in the `PairedUrlStructure` subclasses but also in the `PairedRouting` service.